src/backend/backend.h

```diff
@@ -89,41 +108,92 @@ namespace snmalloc
     static CapPtr<void, CBChunk>
     reserve(GlobalState& h, LocalState* local_state, size_t size)
     {
-      // TODO have two address spaces.
-      UNUSED(is_meta);
+#ifdef SNMALLOC_CHECK_CLIENT
+      constexpr auto MAX_CACHED_SIZE =
+        is_meta ? LOCAL_CACHE_META_BLOCK : LOCAL_CACHE_BLOCK;
+#else
+      constexpr auto MAX_CACHED_SIZE = LOCAL_CACHE_BLOCK;
+#endif
+
+      auto& global = h.address_space;
 
       CapPtr<void, CBChunk> p;
-      if (local_state != nullptr)
+      if ((local_state != nullptr) && (size <= MAX_CACHED_SIZE))
       {
-        p =
-          local_state->local_address_space.template reserve_with_left_over<PAL>(
-            size);
+#ifdef SNMALLOC_CHECK_CLIENT
+        auto& local = is_meta ? local_state->local_meta_address_space :
+                                local_state->local_address_space;
+#else
+        auto& local = local_state->local_address_space;
+#endif
+
+        p = local.template reserve_with_left_over<PAL>(size);
         if (p != nullptr)
-          local_state->local_address_space.template commit_block<PAL>(p, size);
-        else
         {
-          auto& a = h.address_space;
-          // TODO Improve heuristics and params
-          auto refill_size = bits::max(size, bits::one_at_bit(21));
-          auto refill = a.template reserve<false>(refill_size, h.pagemap);
-          if (refill == nullptr)
-            return nullptr;
-          local_state->local_address_space.template add_range<PAL>(
-            refill, refill_size);
-          // This should succeed
-          p = local_state->local_address_space
-                .template reserve_with_left_over<PAL>(size);
-          if (p != nullptr)
-            local_state->local_address_space.template commit_block<PAL>(
-              p, size);
+          return p;
         }
+
+        auto refill =
+          global.template reserve<false>(LOCAL_CACHE_BLOCK, h.pagemap);
+        if (refill == nullptr)
+          return nullptr;
+        auto refill_size = LOCAL_CACHE_BLOCK;
```

Comment:
Maybe consider moving refilll_size above the reservation and using it as the argument, but this is a small nitpick.